src/bmp/encoder.rs

```diff
@@ -33,7 +33,22 @@ impl<'a, W: Write + 'a> BMPEncoder<'a, W> {
         let (dib_header_size, written_pixel_size, palette_color_count) = get_pixel_info(c)?;
         let row_pad_size = (4 - (width * written_pixel_size) % 4) % 4; // each row must be padded to a multiple of 4 bytes
 
-        let image_size = width * height * written_pixel_size + (height * row_pad_size);
+        let image_size = width
+            .checked_mul(height)
+            .ok_or(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Overflowing dimensions",
+            ))?
```

Comment:
Compared to the error code on the `next` branch, this performs some real work and an allocation. Providing these arguments by value means that *all* code paths, not only the erroring ones, will instantiate these three times. This should utilize a closure and `ok_or_else` instead.

```rust
let make_invalid_input_err = || io::Error::new(
    io::ErrorKind::InvalidInput,
    "Overflowing dimensions");

let image_size = width
    .checked_mul(height)
    .ok_or_else(make_invalid_input_err)?
```

Reply:
Great comment thanks. Updated as requested.